File: src/08_Amortized_Analysis/sliding_window_min.cpp
#include <bits/stdc++.h>
#include "../compete.hpp"
using namespace std;


vector<int> v{2, 1, 4, 5, 3, 4, 1, 2};
const size_t n = 8;
const size_t w = 4;
queue<int> q;


void print_min(int i)
{
  cout << "min of window [";
  copy(v.begin() + i - w + 1,
       v.begin() + i + 1,
       ostream_iterator<int>(cout, " "));
  cout << '\b';
  cout << "] with queue " << q << string(2 * w - 2 * q.size(), ' ')
       << " is " << q.front() << '\n';
}


// O(n)
void solve()
{
  cout << v << '\n';
  // first window
  for (size_t i = 0; i < w; ++i) {
    while ((!q.empty()) && (q.front() >= v[i])) q.pop();
    q.push(v[i]);
  }
  print_min(w - 1);

  for (size_t i = w; i < n; ++i) {
    // pop front, if it fell out of the window
    if (q.size() == w) q.pop();

    while((!q.empty()) && (q.front() >= v[i])) q.pop();
    q.push(v[i]);
    print_min(i);
  }
}


int main() {solve();}

File: src/08_Amortized_Analysis/sliding_window_min.py
from collections import deque

v = [2, 1, 4, 5, 3, 4, 1, 2]
w = 4
q = deque(maxlen=w)


def print_min(i):
  window = '[' + ' '.join(map(str, v[i - w + 1: i + 1])) + ']'
  queue = '[' + ' '.join(map(str, q)) + ']'
  print(f"min of window {window} with queue {queue:<{2 * w + 1}} is {q[0]}")


# O(n)
def solve():
  n = len(v)

  if w >= n:
    m = 2e9
    for x in v:
      if x < m:
        m = x
    window = ' '.join(map(str, v))
    queue = ' '.join(map(str, v[v.index(m):]))
    print(f"min of window [{window}] with queue [{queue}] is {m}")
    return

  # first window
  for i in range(w):
    while (q) and (q[0] >= v[i]):
      q.popleft()
    q.append(v[i])

  print_min(w - 1)

  for i in range(w, n):
    if len(q) == w:
      q.popleft()
    while (q) and (q[0] >= v[i]):
      q.popleft()
    q.append(v[i])
    print_min(i)


if __name__ == '__main__':
  solve()
